[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Boyu-Zhang-UOI/pml-f2026-notebooks/blob/main/00-setup/01_session01_taste_demo.ipynb)

# A Whole ML System, End to End

**Session 1 · the in-class taste demo · slides 20–22 carry these same cells**

Ten lines of real machine learning: load 569 tumours, lock a quarter of them
away, fit a model on the rest, and score it — honestly — on the tumours it has
never seen. Nothing here needs to be understood today; by Module 3 every line
will be an old friend.

Runs on the free tier of Colab with nothing to install. Expected numbers are
noted under each cell — small last-digit differences across library versions
are normal.

## Load the data (slide 20)

Rows are examples, columns are features — this shape is the whole semester.
`y` holds the labels: 0 = malignant, 1 = benign.

In [1]:
from sklearn.datasets import load_breast_cancer

data = load_breast_cancer()
X, y = data.data, data.target   # features, labels
print(X.shape, y.shape)
print(data.target_names)
print("malignant:", (y == 0).sum(), " benign:", (y == 1).sum())

(569, 30) (569,)
['malignant' 'benign']
malignant: 212  benign: 357


Remember **212 / 357** — two cells from now that imbalance becomes the
baseline every model has to beat.

## Split, fit, predict (slide 21)

Two ideas in one cell: 143 tumours are locked away *before* training, and the
fitted model then answers about tumours it has never seen.

In [8]:
from sklearn.model_selection import train_test_split  # noqa: I001
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import make_pipeline

i=0
while(i<=9):
    X_tr, X_te, y_tr, y_te = train_test_split(
        X, y, test_size=0.25, random_state=i, stratify=y)
    model = make_pipeline(StandardScaler(),
                          LogisticRegression(max_iter=5000))
    model.fit(X_tr, y_tr)      # "learning" happens here
    print(data.target_names[model.predict(X_te[:1])][0])
    print(f"accuracy on held-out data: {model.score(X_te, y_te):.3f}", "random state: ", i)  #original 0.958, random state 0
    i += 1



benign
accuracy on held-out data: 0.958 random state:  0
malignant
accuracy on held-out data: 0.965 random state:  1
benign
accuracy on held-out data: 0.972 random state:  2
malignant
accuracy on held-out data: 0.986 random state:  3
malignant
accuracy on held-out data: 0.972 random state:  4
benign
accuracy on held-out data: 0.979 random state:  5
benign
accuracy on held-out data: 0.993 random state:  6
benign
accuracy on held-out data: 0.979 random state:  7
benign
accuracy on held-out data: 0.972 random state:  8
benign
accuracy on held-out data: 0.979 random state:  9


## Score, honestly (slide 22)

Always compute the stupid baseline. A classifier that ignores every measurement
and answers "benign" every time is already right 62.9% of the time — so the
thirty measurements bought the distance from 0.629 to 0.958, not from zero.
Train slightly above test is the normal, healthy direction; the size of that
gap becomes a central diagnostic in Module 3.

In [5]:
from sklearn.dummy import DummyClassifier

print(f"train accuracy: {model.score(X_tr, y_tr):.3f}")
print(f"test  accuracy: {model.score(X_te, y_te):.3f}")

dumb = DummyClassifier(strategy="most_frequent").fit(X_tr, y_tr)
print(f"always 'benign': {dumb.score(X_te, y_te):.3f}")

train accuracy: 0.991
test  accuracy: 0.958
always 'benign': 0.629


Expected: `train accuracy: 0.991` · `test accuracy: 0.958` ·
`always 'benign': 0.629`.

That is the whole thing — the six-phase workflow from slide 17, minus deploy,
which waits for Session 29. Slide 23 maps each line above to its phase.

## Where to go next

- **`00_environment_smoke_test.ipynb`** (this folder) — tonight's task: prove
  your own machine can do what this notebook just did.
- **The Session 1 reading**, *The whole thing, once, small* — the same run,
  narrated line by line.